File: sem 3/DS/linked/ll.cpp
#include <iostream>
using namespace std;

class Node {
    int value;
    Node *next;
public:
    Node(int val=0, Node *n=0) {
        value = val;
        next = n;
    }
    friend class LinkedList;
};

class LinkedList {
    Node *head;
public:
    LinkedList() {
        head = 0;
    }

    bool isempty() {
        return head == 0;
    }

    void print() {
        Node *p;
        p = head;
        while (p)
        {
            cout << p->value;
            if (p->next) {
                cout << " -> ";
            }
            p = p->next;
        }
        cout << "\n";
        return;
    }

    void add_start(int val) {
        Node *p;
        p = new Node(val);
        if (head == 0) {
            head = p;
            return;
        } 
        p->next = head;
        head = p;
        return;
    }

    void add_end(int val) {
        Node *p;
        p = new Node(val);
        if (head == 0) {
            head = p;
            return;
        } 
        Node *q;
        q = head;
        while (q->next)
        {
            q = q->next;
        }
        q->next = p;
        return;
    }

    void add_ith(int val, int i) {
        Node *p;
        p = new Node(val);
        if (head == 0) {
            head = p;
            return;
        } 
        Node *q;
        q = head;
        int j = 0;
        while (j<i-1)
        {
            q = q->next;
            j++;
        }
        p->next = q->next;
        q->next = p;
        return;
    }

    void delete_start(){
        Node *p;
        p = head;
        head = head->next;
        delete(p);
        return;
    }

    void delete_end(){
        Node *p;
        p = head;
        while (p->next->next)
        {
            p = p->next;
        }
        Node *q;
        q = p->next;
        p->next = 0;
        delete(q);
        return;
    }

    void delete_ith(int i){
        Node *p;
        p = head;
        int j = 0;
        while (j < i-1)
        {
            p = p->next;
            j++;
        }
        Node *q;
        q = p->next;
        p->next = q->next;
        delete(q);
    }

    void reverse() {
        Node *p;
        Node *q;
        Node *r;
        p = head;
        q = p->next;
        r = q->next;
        p->next = 0;
        while (r)
        {
            q->next = p;
            p = q;
            q = r;
            r = r->next;
        }
        q->next = p;
        head = q;
        return;
    }

    int search(int val) {
        Node *q;
        q = head;
        int j = 0;
        while (q) {
            if (q->value == val) 
                return j;
            q = q->next;
            j++;
        }
        return -1;
        
    }

    void concat(LinkedList ll) {
        Node *q;
        q = head;
        while (q->next)
        {
            q = q->next;
        }
        q->next = ll.head;
        return;
    }

    LinkedList merge(LinkedList ll) {
        LinkedList new_list;
        Node *h1;
        Node *h2;
        h1 = head;
        h2 = ll.head;
        while (h1 && h2) {
            if (h1->value < h2->value) {
                new_list.add_end(h1->value);
                h1 = h1->next;
            } else {
                new_list.add_end(h2->value);
                h2 = h2->next;
            }
        }
        Node *val;
        if (h1) {
            val = h1;
        } else {
            val = h2;
        }
        Node *p;
        while (p->next)
        {
            p = p->next;
        }
        p->next = val;
        return new_list;
    }
};

int main() {
    LinkedList list;
    int nums[10] = {1, 3, 5, 7, 9, 2, 4, 6, 8, 0};
    for (int i=0; i<10; i++) {
        list.add_end(nums[i]);
    }
    list.add_ith(69, 2);

    list.delete_start();
    list.delete_end();
    list.delete_ith(2);
    
    list.print();
    
    list.reverse();
    list.print();
    
    cout << list.search(7) << endl;
    
    LinkedList list2;
    int nums2[5] = {2, 4, 6, 8, 10};
    for (int i = 0; i < 5; i++)
    {
        list2.add_end(nums2[i]);
    }
    list2.print();
    // list.concat(list2);
    // list.print();
    LinkedList new_list;
    new_list = list.merge(list2);
    new_list.print();
    return 0;
}

File: sem 3/DS/linked/ll.py
class Node:
    def __init__(self, value="", next=None):
        self.value = value
        self.next = next
        
class LinkedList:
    def __init__(self):
        self.head = None

    def print(self):
        if self.head:
            p = self.head
            while p:
                print(p.value, end="")
                p = p.next
                if p:
                    print(" -> ", end="")
            print()
    
    def is_empty(self):
        return self.head is None
    
    def add_start(self, val):
        node = Node(val)
        if self.head is None:
            self.head = node
            return
        node.next = self.head
        self.head = node
        
    def add_end(self, val):
        node = Node(val)
        if self.head is None:
            self.head = node
            return
        p = self.head
        while p.next:
            p = p.next
        p.next = node
        
    def add_ith(self, val, i):
        node = Node(val)
        if self.head is None:
            self.head = node
            return
        p = self.head
        j = 0
        while j < i-1:
            p = p.next
            j+=1
        node.next = p.next
        p.next = node
    
    def delete_start(self):
        self.head = self.head.next
    
    def delete_end(self):
        p = self.head
        while p.next.next:
            p = p.next
        p.next = None
        
    def delete_ith(self, i):
        p = self.head
        j = 0
        while j < i-1:
            p = p.next
            j+=1
        q = p.next
        p.next = q.next
        
    def reverse(self):
        p = self.head
        q = p.next
        r = q.next
        p.next = None
        while r:
            q.next = p
            p = q
            q = r
            r = r.next
        q.next = p
        self.head = q
        
    def search(self, val):
        p = self.head
        i = 0
        while p:
            if p.value == val:
                return i
            i+=1
            p = p.next
        return False
    
    def concat(self, llist):
        p = self.head
        while p.next:
            p = p.next
        p.next = llist.head
        
    def merge(self, llist):
        new = LinkedList()
        h1 = self.head
        h2 = llist.head
        while h1 and h2:
            if h1.value > h2.value:
                val = h2
                h2 = h2.next
            else:
                val = h1
                h1 = h1.next
            new.add_end(val.value)
        p = new.head
        while p.next:
            p = p.next

        if h1:
            p.next = h1
        else:
            p.next = h2
            
        return new

if __name__ == "__main__":
    llist = LinkedList()
    for i in [10, 12, 14, 16, 18, 20, 40]:
        llist.add_end(i)
    llist.print()
    # llist.delete_ith(2)
    llist.reverse()
    llist.print()
    
    llist2 = LinkedList()
    for i in [1, 3, 5, 7, 9]:
        llist2.add_end(i)
    new = llist.merge(llist2)
    new.print()